### PR TITLE
[Feature] Event for Requested Theme Changes

### PR DESCRIPTION
**Pull request by jamesmontemagno**
**NEEDS REVIEW**

### Description of Change ###

Enable developers to subscribe to app theme request changes. Also add in platform helpers on iOS/Android to bubble up platform events.

### API Changes ###
App Info:
* Added: public static EventHandler<AppThemeChangedEventArgs> RequestedThemeChanged
* Added: AppThemeChangedEventArgs

In Platform files:
* Added on iOS & tvOS: public static void TraitCollectionDidChange(UITraitCollection currentUITraitCollection)
* Added on Android:  public static void OnConfigurationChanged(Configuration newConfig)

### Behavioral Changes ###

None.

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Has samples (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [ ] Updated documentation ([see walkthrough](https://github.com/xamarin/Essentials/wiki/Documenting-your-code-with-mdoc))

### CODE REVIEW

**Comment by jamesmontemagno** [comment]
I would also like @PureWeen to take a look at this or any other Xamarin.Forms teams members.

My thought is that Xamarin.Forms could automatically register for the underlying events so the developer does need to do it.

**Comment by PureWeen** [comment]
@jfversluis

**Comment by jfversluis** [comment]
So, this is basically how I have implemented it at some point. My consideration for Android was this. For the Android (auto) dark theme colors, to kick in you will need to recreate the activity. If you do not handle the ConfigurationChanged event this will happen automatically, if you handle the ConfigurationChanged like you are here then you will need to trigger the recreation yourself. By doing that though, you _can_ fire your own event, but that will have very limited use because the activity is recreated right after you did whatever you want to do in our event, right? Does that make sense?

There are of course other use cases and I guess this then makes sense and Forms can hook into this. So from that standpoint this looks fine :)

**Comment by jamesmontemagno** [comment]
So, if you have the default:

```
[Activity(
    // other settings
    ConfigurationChanges = ConfigChanges.ScreenSize | ConfigChanges.Orientation)]
```

Then your application activity is killed and re-inflated, which is really bad and potentially will crash your application. 

If you add
```
| ConfigChanges.UiMode)]
```
Then your activity will remain, your state will remain and you can update the UI. This is how all templates should be setup. 

So, the thought here is that a "FormssAppCompatAcitivty" could forward the event to Essentials to automatically trigger the event handler. 

This means that all a user would have to do is add:
```
| ConfigChanges.UiMode)]
```
and not have to override the event themselves. 

Ideally Xamarin.Forms could also Init essentials, do the permissions override too :) but that is for another conversation.

**Comment by jfversluis** [comment]
Totally on board with how this works, but for the Android OS theme to actually flip in your app when you do the `| ConfigChanges.UiMode)]`, you will now need to manually call `Recreate();`, that is what I understand from it anyway. Hopefully I'm missing something obvious here.

So, if that's something you are ignoring as a developer and do all the theming yourself then that is fine. But if you want to lean into the automatic theme done by the DayNight theme you will still have to recreate the activity. See attached project and screen recording to hopefully get my point across. 

Maybe I'm focussing too much on a use case that is not very common or we are willing to document and get over with :)

[AndroidConfigTest.zip](https://github.com/xamarin/Essentials/files/4372045/AndroidConfigTest.zip)

https://1drv.ms/v/s!Anrk6U9nRknzg58bMYOsBj6dOqcalQ?e=6aUL73

**Comment by jamesmontemagno** [comment]
I mean I just "apply theme" when this happens: https://github.com/jamesmontemagno/Hanselman.Forms

All my styles are dynamic. I think we can document that you may need to re-create, but ideally your app just stays in the same place and not re-boot the entire activity. We can hop on a call to discuss though :)

I think Android will just update the UI if you are using the correct theme: https://developer.android.com/guide/topics/ui/look-and-feel/darktheme ?

**Comment by jfversluis** [comment]
Yeah, so, I guess that makes sense. Most likely scenarios are; you take full control over your theming and use this. Or use the automatic theming and do not worry about the event and just let the recreate do its thing.

Ultimately it just comes down to documenting this well and let people make their own choice.

File: Xamarin.Essentials/AppInfo/AppInfo.uwp.cs [hunk 1]
```diff
@@ -24,5 +30,19 @@ static string PlatformGetVersionString()
...
+
+        static void StartThemeListeners() =>
+            UISettings.ColorValuesChanged += UISettingsColorValuesChanged;
+
+        static void StopThemeListeners() =>
+            UISettings.ColorValuesChanged -= UISettingsColorValuesChanged;
+
+        static void UISettingsColorValuesChanged(UISettings sender, object args)
+        {
+            MainThread.BeginInvokeOnMainThread(() =>
```
Comment (Mrnikbobjeff) [inline Xamarin.Essentials/AppInfo/AppInfo.uwp.cs:42]: Is this needed? We already switch to the mainthread in the OnAppThemeChanged

**Comment by mattleibow** [comment]
I am closing this not because this is not good to have, but with the new way with MAUI, this is better served in that project. Also, with the merging of things, this is better to be re-investigated in the new world.

**Pull request closed without merging**